splay/src/main/scala/org/scalajs/benchmark/splay/Splay.scala

```diff
@@ -0,0 +1,439 @@
+          // root.
+          root.right = right
+        }
+        removed
+      }
+    }
+  }
+
+  /**
+   * Returns the node having the specified key or null if the tree doesn't contain
+   * a node with the specified key.
+   *
+   * @param {number} key Key to find in the tree.
+   * @return {SplayTree.Node} Node having the specified key.
+   */
+  def find(key: Int): Node = {
+    if (isEmpty) {
+      null
+    } else {
+      splay(key)
+      if (root.key == key) root
+      else null
+    }
+  }
+
+  /**
+   * @return {SplayTree.Node} Node having the maximum key value.
+   */
+  def findMax(opt_startNode: Node): Node = {
+    if (isEmpty) {
+      null
+    } else {
+      var current =
+        if (opt_startNode != null) opt_startNode
+        else root
+      while (current.right != null) {
+        current = current.right
+      }
+      current
+    }
+  }
+
+  /**
+   * @return {SplayTree.Node} Node having the maximum key value that
+   *     is less than the specified key value.
+   */
+  def findGreatestLessThan(key: Int): Node = {
+    if (isEmpty) {
+      null
+    } else {
+      // Splay on the key to move the node with the given key or the last
+      // node on the search path to the top of the tree.
+      splay(key)
+      // Now the result is either the root node or the greatest node in
+      // the left subtree.
+      if (root.key < key) {
+        root
+      } else if (root.left != null) {
+        findMax(root.left)
+      } else {
+        null
+      }
+    }
+  }
+
+  /**
+   * @return {Array<*>} An array containing all the keys of tree's nodes.
+   */
+  def exportKeys: Array[Int] = {
+    // Both variants have a similar performance
+     var b = Array.newBuilder[Int]
+     if (!isEmpty)
+        root.traverse(node => b += node.key)
+     b.result()
+//    val arr = new js.Array[Int]
+//    if (!isEmpty)
+//      root.traverse(node => arr.push(node.key))
+//    arr
+  }
+
+  /**
+   * Perform the splay operation for the given key. Moves the node with
+   * the given key to the top of the tree.  If no node has the given
+   * key, the last node on the search path is moved to the top of the
+   * tree. This is the simplified top-down splaying algorithm from:
+   * "Self-adjusting Binary Search Trees" by Sleator and Tarjan
+   *
+   * @param {number} key Key to splay the tree on.
+   * @private
+   */
+  private def splay(key: Int): Unit = {
+    if (isEmpty) {
+      // Do nothing
+    } else {
+      // Create a dummy node.  The use of the dummy node is a bit
+      // counter-intuitive: The right child of the dummy node will hold
+      // the L tree of the algorithm.  The left child of the dummy node
+      // will hold the R tree of the algorithm.  Using a dummy node, left
+      // and right will always be nodes and we avoid special cases.
+      val dummy = Node.newDummy()
+      var left = dummy
+      var right = dummy
+      var current = root
+      var break = false
+      while (!break) {
```

Comment:
It would be more idiomatic and (probably) more performance to use an `@tailrec` inner `def` here, instead of a loop with weird `break`s.